lib/active_merchant/billing/gateways/ct_payment.rb

```diff
@@ -0,0 +1,258 @@
+        }
+        add_operator_id(post, options)
+        add_payment(post, credit_card)
+        add_customer_data(post, options)
+
+        commit('recur/AddUser', post)
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((&?auth-api-key=)[^&]*)i, '\1[FILTERED]').
+          gsub(%r((&?payload=)[a-zA-Z%0-9=]+)i, '\1[FILTERED]').
+          gsub(%r((&?token:)[^&]*)i, '\1[FILTERED]').
+          gsub(%r((&?cardNumber:)[^&]*)i, '\1[FILTERED]')
+      end
+
+      private
+
+      def add_terminal_number(post, options)
+        post[:MerchantTerminalNumber] = options[:merchant_terminal_number] || ' ' * 5
+      end
+
+      def add_money(post, money)
+        post[:Amount] = money.to_s.rjust(11,'0')
+      end
+
+      def add_operator_id(post, options)
+        post[:OperatorID] = options[:operator_id] || '0' * 8
+      end
+
+      def add_customer_data(post, options)
+        post[:CustomerNumber] = options[:customer_number] || '0' * 8
+      end
+
+      def add_address(post, creditcard, options)
+        if address = options[:billing_address] || options[:address]
+          post[:CardHolderAddress] = ("#{address[:address1]} #{address[:address2]}").rjust(20, ' ')
+          post[:CardHolderPostalCode] = address[:zip].gsub(/\s+/, '').rjust(9, ' ')
+        end
+      end
+
+      def add_invoice(post, money,  options)
+        post[:CurrencyCode] = options[:currency] || (currency(money) if money)
+        post[:InvoiceNumber] = options[:order_id].rjust(12,'0')
+        post[:InputType] = 'I'
+        post[:LanguageCode] = 'E'
+      end
+
+      def add_payment(post, payment)
+        if payment.is_a?(String)
+          post[:Token] = split_authorization(payment)[3].strip
+        else
+          post[:CardType] = CARD_BRAND[payment.brand] || ' '
+          post[:CardNumber] = payment.number.rjust(40,' ')
+          post[:ExpirationDate] = expdate(payment)
+          post[:Cvv2Cvc2Number] = payment.verification_value
+        end
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def split_authorization(authorization)
+        authorization.split(';')
+      end
+
+      def commit(action, parameters, send_ack = true)
+        url = (test? ? test_url : live_url) + action
+        response = parse(ssl_post(url, post_data(action, parameters)))
+
+        final_response = Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          avs_result: AVSResult.new(code: response['avsStatus']),
+          cvv_result: CVVResult.new(response['cvv2Cvc2Status']),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+        if send_ack
+          post = {
+            TransactionNumber: split_authorization(final_response.authorization)[0]
+          }
+          commit('ack', post, false)
```

Comment:
Hmm... did you consider using a MultiResponse for this instead?

Reply:
I did... trying to remember why I went back to this way. 🤔

Reply:
@curiousepic for the MultiResponse wouldn't it need to be on the first commit and not nested inside of it?

Reply:
Yep. Could be part of the commit method or it could be part of the actions, ala an auth+capture purchase. I'm honestly not sure if it's a good idea yet, but it seems like a pattern that already exists that we're recreating from scratch here. There could be a qualitative difference I can't see, though.

Reply:
Yea I had in each method, but this reduced lines of code.

Reply:
> @curiousepic for the MultiResponse wouldn't it need to be on the first commit and not nested inside of it?

This is moot at this point, but I'm confused. If I get what you're saying, you can control that via whether you pass `use_first_response` as `true` or not to `MultiResponse.run`, so that'd just be a slightly different approach than this; you'd end up with something like

```ruby
def commit_raw(action, parameters)
  ...current contents of commit method...
end

def commit(action, parameters)
  MultiResponse.run(true) do |r|
    resp = r.commit_raw(action, parameters)
    r.commit_raw('ack', {TransactionNumber: split_authorization(resp.authorization)[0]})
  end
end
```

(To be clear, I'm not asking you change what you're doing here; just not clear on why `MultiResponse` wouldn't do the trick.)

Reply:
@bpollack I as more so asking how could it be done with the current commit structure. Definitely see how add a new commit `commit_raw` could achieve the goal though.

Reply:
@bpollack and @curiousepic new and improved with `MultiResponse`